natss/pkg/reconciler/dispatcher/natsschannel_test.go

```
@@ -0,0 +1,202 @@
+	ncName = "test-nc"
+)
+
+func TestAllCases(t *testing.T) {
+	ncKey := testNS + "/" + ncName
+
+	table := TableTest{
+		{
+			Name: "bad workqueue key",
+			// Make sure Reconcile handles bad keys.
+			Key: "too/many/parts",
+		},
+		{
+			Name: "key not found",
+			// Make sure Reconcile handles good keys that don't exist.
+			Key: "foo/not-found",
+		},
+		{
+			Name: "reconcile ok: channel ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					reconciletesting.WithReady,
+				),
+			},
+			WantPatches: []clientgotesting.PatchActionImpl{
+				makeFinalizerPatch(testNS, ncName),
+			},
+			WantErr: false,
+		},
+		{
+			Name: "reconcile error: channel not ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					reconciletesting.WithNotReady("ups", ""),
+				),
+			},
+			WantErr: true,
+		},
+		{
+			Name: "remove finalizer even though channel is not ready",
+			Key:  ncKey,
+			Objects: []runtime.Object{
+				reconciletesting.NewNatssChannel(ncName, testNS,
+					// the finalizer should get removed
+					reconciletesting.WithNatssChannelFinalizer,
+					// make sure that finalizer can get removed even if channel is not ready
+					reconciletesting.WithNotReady("ups", ""),
+					reconciletesting.WithNatssChannelDeleted),
+			},
+			WantErr: false,
+			WantUpdates: []clientgotesting.UpdateActionImpl{
+				{
+					Object: reconciletesting.NewNatssChannel(ncName, testNS,
+						// finalizer is removed
+						// make sure that finalizer can get removed even if channel is not ready
+						reconciletesting.WithNotReady("ups", ""),
+						reconciletesting.WithNatssChannelDeleted),
+				},
+			},
+		},
+	}
+	defer logtesting.ClearAll()
+
+	table.Test(t, reconciletesting.MakeFactory(func(listers *reconciletesting.Listers, opt reconciler.Options) controller.Reconciler {
+		natssDispatcher := dispatchertesting.NewDispatcherDoNothing(t)
```

Do we really want to default to a mock? I believe it would be more reliable to provide a mocked channel than a mocked dispatcher.

---

Can you elaborate more? Dispatcher has a connection to natss, so we would need to run natss in the test. Is your idea to make the test an integration test?

---

It's not an integration test if you mock NATSS. Anything that satisfies the NATSS API used by the dispatcher is fine. I just don't see the point mocking the thing that you test (the dispatcher), mocks are for dependencies.

---

As discussed in person: reconciler has a dispatcher and the object under test is the reconciler, not the dispatcher.